Handle 'uses' delegating to another 'uses' that has a deep hash

kbrock opened the pull request:
## Issue

The use case that triggered this had more layers of indirection.
But the jist is when we map a `:uses` with an array, and one of the entries has a complex `:uses`,
it will end up with a nested hash.

```
virtual_has_many :field1, :uses => {:a => :b}
virtual_has_many :field2, :uses => :c
virtual_has_many :field3, :uses => [:field1, :field2]
```

## Before

```
references([:field3])
=>
references([:field1, :field2])
=>
references([{:a => :b}, :c]
```

And when we map that into a hash we end up with
```
{{:a => :b} => {}, :c => {}}
```
where `{:a => :b}` is a key pointing to an empty hash (i.e.: `{}`)

### So we blow up

```
Issue with `Vm#used_storage`:

ERROR -- evm: [NoMethodError]: undefined method `to_sym' for {:hardware=>:disks}:Hash
Did you mean?  to_s
               to_set
ERROR -- evm: lib/active_record/virtual_attributes/virtual_reflections.rb:36:in `virtual_reflection?`
lib/active_record/virtual_attributes/virtual_fields.rb:30:in `virtual_field?`
lib/active_record/virtual_attributes/virtual_fields.rb:50:in `block in replace_virtual_field_hash`
lib/active_record/virtual_attributes/virtual_fields.rb:49:in `each`
```


## After

We don't assume that all entries in the input array are a string/symbol:
so we end up with simple keys in the destination hash:

```
{:a => :b, :c => {}}
```

Fixes https://github.com/ManageIQ/manageiq/issues/22526

kbrock commented:
update:
- fixed spacing cops
- added another test case and better document existing test case. (I'm concerned that future me will make the code more efficient and basically remove the case that we are testing)

Fryguy commented on spec/db/models.rb:
```diff
@@ -95,6 +99,11 @@ def book_with_most_bookmarks
   # uses another virtual attribute that uses a relation
   virtual_attribute :upper_first_book_author_name, :string, :uses => :first_book_author_name
+  # uses another virtual_attribute that uses a complex relation
+  # Please do not "fix" syntax of the uses here
+  #   actually testing this format with an array: [:bwmb, {:books => co_a}]
+  #   vs "better" format not helping us with tests: {:bwmb => {}, :books => co_a}
```
The wording here is strange to me - I think this is clearer and more direct

```suggestion
  # NOTE: Please do not change the :uses format here.
  #   This intentionally tests :uses with an array: [:bwmb, {:books => co_a}]
  #   vs a "better" format such as: {:bwmb => {}, :books => co_a}
```

Fryguy commented on spec/virtual_includes_spec.rb:
```diff
@@ -462,6 +487,24 @@
   end
 
+  context ".replace_virtual_field (private)" do
+    # would have prefer target: {:books => :bookmarks} without the extra :books. but here we are
```
I'm not sure what this comment means.

kbrock replied:
it was trying to say that an extra `:books` comes out.
And that is not optimal but is good enough. Pretty sure rails ignores.

Thanks. Rewrote as a TODO.

Fryguy commented on spec/virtual_includes_spec.rb:
```diff
@@ -462,6 +487,24 @@
+      expect(Author.replace_virtual_fields([{:book_with_most_bookmarks => {}}, {:books => {}}])).to eq([{:books => :bookmarks}, {:books => {}}])
+    end
+
+    # I was a little confused that it uses author (association) instead of authors (table name)
+    # but active record seems ok with this
```
This is weird in code to me - it's like a self-narrative conversation?  Can we write this in third person or something for other readers?

```suggestion
    # TODO: Need to understand why this uses author (association) instead of authors (table name)
    # and why active record seems ok with this
```

kbrock replied:
yea. Think it was just me grumbling. I deleted

Fryguy commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -69,16 +69,23 @@ def include_to_hash(value)
             {value => {}}
           when Array
-            value.flatten.each_with_object({}) { |k, h| h[k] = {} }
+            value.flatten.reduce({}) do |h, k|
+              if k.kind_of?(Hash)
+                merge_includes(h, k)
+              else
+                h[k] ||= {}
+                h
+              end
+            end
```
Super-minor, but I'm thikning maybe we shouldn't rely on merge_includes returning the hash?  Then we can DRY up the "return the h hash" part

```suggestion
            value.flatten.reduce({}) do |h, k|
              if k.kind_of?(Hash)
                merge_includes(h, k)
              else
                h[k] ||= {}
              end
              h
            end
```

kbrock replied:
made that change. it squashed it down even further

Fryguy reviewed (approved):
Overall LGTM - My only comments are very minor and/or comments about the comments.

I'll hold off merge based on how you want to move forward and when you want to release the gem with this fix.

kbrock commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -69,16 +69,23 @@ def include_to_hash(value)
 
-        # @param [Hash] hash1
-        # @param [Hash] hash2
+        # @param [Hash] hash1 (typically modified)
+        # @param [Hash] hash2 (not modified)
```
this is private.

kbrock replied:
the actual code has changed to always modified / not modified

kbrock commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -69,23 +69,24 @@ def include_to_hash(value)
             {value => {}}
           when Array
-            value.flatten.each_with_object({}) { |k, h| h[k] = {} }
+            value.flatten.each_with_object({}) do |k, h|
+              merge_includes(h, k)
+            end
```
If we tread the string/symbol case as a special case, it would avoid converting `h` to a hash `h => {}` and avoid the deep_merge() call. But I think what we have is better overall

```suggestion
            value.flatten.each_with_object({}) do |k, h|
              if k.kind_of?(Symbol) || k.kind_of?(String)
                h[k] ||= {}
              else
                merge_includes(h, k)
              end
            end
```

kbrock replied:
This case is pretty common.

I added it to the code so it wasn't in the back of my mind.

kbrock replied:
I implemented this behavior.
If I cared enough about this "regression" to document it here, then I cared enough about it to write up some tests and fix it.

kbrock commented:
@Fryguy you good with this one?

kbrock commented:
update:

- ~~symbolize association names so they merge better (most likely not necessary)~~
- added special case/optimization when you are 
- added tests around `:includes => [:virtual_attribute_with_no_includes]`

update:

- pulled all the `to_sym` changes into a separate PR

kbrock commented:
update:
- brought in the tests that were in the extracted PR (for `to_sym`)

kbrock commented:
@Fryguy This should be good to go.

I pulled out anything that was not directly related to this PR.

I did put in a ton of edge cases (that are probably a little over kill)

Note: Fixes https://github.com/ManageIQ/manageiq/issues/22526
